subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -487,6 +518,15 @@ public boolean executeWithTask(WorkerLease parentWorkerLease, final Action<TaskI
                     }
                 }
 
+                workRemaining.set(workRemaining());
+                if (!workRemaining.get()) {
+                    return FINISHED;
+                }
+
+                if (allProjectsLocked()) {
```

Comment:
That's the thing with these kinds of changes. Took me weeks of incubating your ideas and days of coding myself and then the result is something where everybody would say "Duh!".